examples/tv-casting-app/linux/main.cpp

```diff
@@ -237,23 +227,280 @@ void OnContentLauncherFailureResponse(void * context, CHIP_ERROR error)
+        {
+            return endpoint;
+        }
+        for (auto & endpointInfo : mEndpoints)
+        {
+            if (!endpointInfo.IsInitialized())
+            {
+                endpointInfo.Initialize(endpointId);
+                return &endpointInfo;
+            }
+        }
+        return nullptr;
+    }
+
+    TargetEndpointInfo * GetEndpoint(EndpointId endpointId)
+    {
+        if (!mInitialized)
+        {
+            return nullptr;
+        }
+        for (auto & endpointInfo : mEndpoints)
+        {
+            if (endpointInfo.IsInitialized() && endpointInfo.GetEndpointId() == endpointId)
+            {
+                return &endpointInfo;
+            }
+        }
+        return nullptr;
+    }
+
+    bool HasEndpoint(EndpointId endpointId)
+    {
+        if (!mInitialized)
+        {
+            return false;
+        }
+        for (auto & endpointInfo : mEndpoints)
+        {
+            if (endpointInfo.IsInitialized() && endpointInfo.GetEndpointId() == endpointId)
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    void PrintInfo()
+    {
+        ChipLogProgress(NotSpecified, " TargetVideoPlayerInfo nodeId=0x" ChipLogFormatX64 " fabric index=%d",
+                        ChipLogValueX64(mNodeId), mFabricIndex);
+        for (auto & endpointInfo : mEndpoints)
+        {
+            if (endpointInfo.IsInitialized())
+            {
+                endpointInfo.PrintInfo();
+            }
+        }
+    }
+
+private:
+    static constexpr size_t kMaxNumberOfEndpoints = 5;
+    TargetEndpointInfo mEndpoints[kMaxNumberOfEndpoints];
+    NodeId mNodeId;
+    chip::FabricIndex mFabricIndex;
+    bool mInitialized = false;
+};
+TargetVideoPlayerInfo gTargetVideoPlayerInfo;
+
+void OnDescriptorReadSuccessResponse(void * context, const chip::app::DataModel::DecodableList<chip::ClusterId> & responseList)
+{
+    TargetEndpointInfo * endpointInfo = static_cast<TargetEndpointInfo *>(context);
+
+    ChipLogProgress(AppServer, "Descriptor: Default Success Response endpoint=%d", endpointInfo->GetEndpointId());
+
+    auto iter = responseList.begin();
+    while (iter.Next())
+    {
+        auto & clusterId = iter.GetValue();
+        endpointInfo->AddCluster(clusterId);
+    }
+    gTargetVideoPlayerInfo.PrintInfo();
+}
+
+void OnDescriptorReadFailureResponse(void * context, CHIP_ERROR error)
+{
+    ChipLogError(AppServer, "Descriptor: Default Failure Response: %" CHIP_ERROR_FORMAT, error.Format());
+}
+
+void ReadServerClusters(EndpointId endpointId)
+{
+    Server * server           = &(chip::Server::GetInstance());
+    chip::FabricInfo * fabric = server->GetFabricTable().FindFabricWithIndex(gTargetVideoPlayerInfo.GetFabricIndex());
+    if (fabric == nullptr)
+    {
+        ChipLogError(AppServer, "Did not find fabric for index %d", gTargetVideoPlayerInfo.GetFabricIndex());
+        return;
+    }
+
+    chip::DeviceProxyInitParams initParams = {
```

Comment:
Why are we doing this by hand instead of using the `CASESessionManager` hanging off `Server`?

Reply:
for some reason, calling FindExistingSession on the server's CASESessionManager is not working

Reply:
Ah, I had missed that this was pre-existing code.  We really need to fix it to do the right thing, but yes, not in this PR.